messaging/src/main/java/org/axonframework/eventhandling/deadletter/jpa/JpaSequencedDeadLetterQueue.java

```diff
@@ -0,0 +1,716 @@
+                .createQuery(
+                        "SELECT dl FROM DeadLetterEntry dl where dl.processingGroup=:processingGroup and dl.sequence=:identifier",
+                        DeadLetterEntry.class)
+                .setParameter("processingGroup", processingGroup)
+                .setParameter("identifier", identifier)
+                .getResultList()
+                .stream()
+                .map(this::toLetter)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public Iterable<Iterable<DeadLetter<? extends M>>> deadLetters() {
+        List<String> sequences = entityManagerProvider
+                .getEntityManager()
+                .createQuery(
+                        "SELECT dl.sequence FROM DeadLetterEntry dl where dl.processingGroup=:processingGroup order by dl.lastTouched asc",
+                        String.class)
+                .setParameter("processingGroup", processingGroup)
+                .getResultList();
+
+
+        return () -> {
+            Iterator<String> sequenceIterator = sequences.iterator();
+            return new Iterator<Iterable<DeadLetter<? extends M>>>() {
+                @Override
+                public boolean hasNext() {
+                    return sequenceIterator.hasNext();
+                }
+
+                @Override
+                public Iterable<DeadLetter<? extends M>> next() {
+                    String next = sequenceIterator.next();
+                    return deadLetterSequence(next);
+                }
+            };
+        };
+    }
+
+    /**
+     * Converts a {@link DeadLetterEntry} from the database into a {@link JpaDeadLetter}, using the configured
+     * {@link DeadLetterJpaConverter DeadLetterJpaConverters} to restore the original message from it.
+     *
+     * @param entry The entry to convert.
+     * @return The {@link DeadLetter} result.
+     */
+    @SuppressWarnings("unchecked")
+    private JpaDeadLetter<M> toLetter(DeadLetterEntry entry) {
+        MetaData deserializedDiagnostics = serializer.deserialize(new SimpleSerializedObject<>(
+                entry.getDiagnostics(),
+                byte[].class,
+                MetaData.class.getName(),
+                null));
+        DeadLetterJpaConverter<M> converter = (DeadLetterJpaConverter<M>) converters
+                .stream()
+                .filter(c -> c.canConvert(entry.getMessage()))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("No such converter"));
+        return new JpaDeadLetter<>(entry,
+                                   deserializedDiagnostics,
+                                   converter.fromEntry(entry.getMessage(), serializer));
+    }
+
+    @Override
+    public boolean isFull(@Nonnull Object sequenceIdentifier) {
+        String identifier = toIdentifier(sequenceIdentifier);
+        return maximumNumberOfQueuesReached(identifier) || maximumQueueSizeReached(identifier);
```

Comment:
It seems like you're retrieving the size of a sequence twice now. Can't we optimize that by retrieve a given sequence size just once?